File: Dead-by-Daylight-Server-Fix.Core/HostsFileService.cs
using System.Text;

namespace SelectRegionForDbd.Core;

public static class HostsFileService
{
    private const string HostsPath = @"C:\Windows\System32\drivers\etc\hosts";

    private static readonly string[] SectionHeaders = ["#asia", "#america", "#europe"];

    public static OpResult Apply(GameliftRegion selected)
    {
        try
        {
            var lines = File.Exists(HostsPath)
                ? File.ReadAllLines(HostsPath).Where(line => !IsOwnedLine(line)).ToList()
                : [];

            lines.AddRange(BuildBlock(selected));
            File.WriteAllLines(HostsPath, lines);
            DnsUtil.Flush();
            return OpResult.Ok;
        }
        catch (UnauthorizedAccessException)
        {
            return OpResult.Fail("You do not have permission to modify the hosts file. Run the application as administrator.");
        }
        catch (Exception ex)
        {
            return OpResult.Fail($"An error occurred: {ex.Message}");
        }
    }

    public static OpResult Clear()
    {
        try
        {
            if (!File.Exists(HostsPath))
            {
                return OpResult.Ok;
            }

            var lines = File.ReadAllLines(HostsPath).Where(line => !IsOwnedLine(line)).ToArray();
            File.WriteAllLines(HostsPath, lines);
            DnsUtil.Flush();
            return OpResult.Ok;
        }
        catch (UnauthorizedAccessException)
        {
            return OpResult.Fail("You do not have permission to modify the hosts file. Run the application as administrator.");
        }
        catch (Exception ex)
        {
            return OpResult.Fail($"An error occurred: {ex.Message}");
        }
    }

    // Применённая зона определяется по закомментированной (разрешённой) строке в hosts
    public static string? GetAppliedRegionId()
    {
        try
        {
            if (!File.Exists(HostsPath))
            {
                return null;
            }

            var commentedLines = File.ReadAllLines(HostsPath)
                .Select(line => line.Trim())
                .Where(line => line.StartsWith('#'))
                .ToArray();

            foreach (var region in RegionCatalog.All)
            {
                if (commentedLines.Any(line => line.Contains(region.Host, StringComparison.OrdinalIgnoreCase)))
                {
                    return region.Id;
                }
            }
            return null;
        }
        catch
        {
            return null;
        }
    }

    private static bool IsOwnedLine(string line)
    {
        string trimmed = line.Trim();
        return trimmed is "#asia" or "#america" or "#europe" ||
               (trimmed.Contains("gamelift.", StringComparison.OrdinalIgnoreCase) &&
                trimmed.Contains("amazonaws.com", StringComparison.OrdinalIgnoreCase));
    }

    private static IEnumerable<string> BuildBlock(GameliftRegion selected)
    {
        var block = new List<string>();
        foreach (string group in SectionHeaders)
        {
            block.Add(group);
            foreach (var region in RegionCatalog.HostsOrder.Where(r => r.HostsGroup == group))
            {
                bool commented = region.Id == selected.Id;
                string prefix = commented ? "#0.0.0.0" : "0.0.0.0";
                block.Add($"{prefix} {region.Host} # {region.Name} {region.Id}");
            }
        }
        return block;
    }
}

File: src/core/HostsFileService.cpp
#include "HostsFileService.h"

#include <windows.h>

#include "DnsUtil.h"
#include "Region.h"
#include "RegionCatalog.h"
#include "Util.h"

namespace HostsFileService {

static const std::string HostsPath = "C:\\Windows\\System32\\drivers\\etc\\hosts";

static bool IsOwnedLine(const std::string& line)
{
    std::string t = util::trim(line);
    if (t == "#asia" || t == "#america" || t == "#europe")
        return true;
    return util::containsIgnoreCase(t, "gamelift.") && util::containsIgnoreCase(t, "amazonaws.com");
}

static void BuildBlock(const Region& selected, std::vector<std::string>& block)
{
    static const char* groups[] = {"#asia", "#america", "#europe"};
    for (const char* group : groups)
    {
        block.push_back(group);
        for (const auto& region : RegionCatalog::All())
        {
            if (region.hostsGroup != group)
                continue;
            bool commented = (region.id == selected.id);
            std::string prefix = commented ? "#0.0.0.0" : "0.0.0.0";
            block.push_back(prefix + " " + region.host + " # " + region.name + " " + region.id);
        }
    }
}

Result Apply(const Region& selected)
{
    try
    {
        std::vector<std::string> lines;
        if (util::fileExists(HostsPath))
        {
            bool ok = false;
            auto all = util::readAllLines(HostsPath, ok);
            if (ok)
            {
                for (const auto& line : all)
                {
                    if (!IsOwnedLine(line))
                        lines.push_back(line);
                }
            }
        }
        BuildBlock(selected, lines);
        if (!util::writeAllLines(HostsPath, lines))
        {
            return {false, "You do not have permission to modify the hosts file. Run the application as administrator."};
        }
        DnsUtil::Flush();
        return {true, ""};
    }
    catch (...)
    {
        return {false, "An error occurred while modifying the hosts file."};
    }
}

Result Clear()
{
    try
    {
        if (!util::fileExists(HostsPath))
            return {true, ""};

        bool ok = false;
        auto all = util::readAllLines(HostsPath, ok);
        if (!ok)
            return {false, "Could not read the hosts file."};

        std::vector<std::string> cleaned;
        for (const auto& line : all)
        {
            if (!IsOwnedLine(line))
                cleaned.push_back(line);
        }

        if (!util::writeAllLines(HostsPath, cleaned))
        {
            return {false, "You do not have permission to modify the hosts file. Run the application as administrator."};
        }
        DnsUtil::Flush();
        return {true, ""};
    }
    catch (...)
    {
        return {false, "An error occurred while modifying the hosts file."};
    }
}

std::string GetAppliedRegionId()
{
    try
    {
        if (!util::fileExists(HostsPath))
            return "";

        bool ok = false;
        auto lines = util::readAllLines(HostsPath, ok);
        if (!ok)
            return "";

        std::vector<std::string> commented;
        for (const auto& line : lines)
        {
            std::string t = util::trim(line);
            if (!t.empty() && t[0] == '#')
                commented.push_back(t);
        }

        for (const auto& region : RegionCatalog::All())
        {
            for (const auto& c : commented)
            {
                if (util::containsIgnoreCase(c, region.host))
                    return region.id;
            }
        }
        return "";
    }
    catch (...)
    {
        return "";
    }
}

} // namespace HostsFileService